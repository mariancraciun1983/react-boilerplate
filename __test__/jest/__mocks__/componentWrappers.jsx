import React  from 'react';
import clone from 'clone-deep';
import configureStore from 'redux-mock-store';
import { BrowserRouter as Router } from 'react-router-dom';
import { Provider } from 'react-redux';

import mockStore  from './store';


const { fullStore } = mockStore;

const genericWrapper = (children) => {
  const mockStore = configureStore();
  const initialState = clone(fullStore);
  const store = mockStore(initialState);
  return (
    <Provider store={store}>
      <Router>
        {children}
      </Router>
    </Provider>
  )
};

export {
  genericWrapper
};
