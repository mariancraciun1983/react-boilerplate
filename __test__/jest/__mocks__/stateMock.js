export default {
  app: {
    state: "loaded"
  },
  auth: {
    authenticated: true,
    token: "abc123",
    user: {
      name: "John Doe",
      email: "john@example.com",
      avatar: "https://dummyimage.com/100x100.png/5fa2d1/ffffff"
    }
  },
  genres: {
    list: [
      {
        name: "Comedy",
        slug: "comedy"
      },
      {
        name: "Adventure",
        slug: "adventure"
      }
    ]
  },
  movies: {
    list: [
      {
        id: "0b7b8b05-6c88-4517-a969-adf89a46bc0a",
        title: "Lost Palace",
        genre: ["comedy"],
        image: "http://dummyimage.com/100x100.png/cc0000/ffffff",
        description:
          "Donec diam neque, vestibulum eget, vulputate ut, ultrices vel, augue. Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia Curae; Donec pharetra, magna vestibulum aliquet ultrices, erat tortor sollicitudin mi, sit amet lobortis sapien sapien non mi. Integer ac neque.\n\nDuis bibendum. Morbi non quam nec dui luctus rutrum. Nulla tellus.\n\nIn sagittis dui vel nisl. Duis ac nibh. Fusce lacus purus, aliquet at, feugiat non, pretium quis, lectus.\n\nSuspendisse potenti. In eleifend quam a odio. In hac habitasse platea dictumst.\n\nMaecenas ut massa quis augue luctus tincidunt. Nulla mollis molestie lorem. Quisque ut erat.",
        price: 2.39,
        slug: "gunner-palace",
        title_lower: "lost palace"
      },
      {
        id: "f01187b5-6fe8-465b-a0b8-3764a94ae225",
        title: "Love is Eternal While It Lasts",
        genre: ["comedy", "romance"],
        image: "http://dummyimage.com/100x100.png/5fa2dd/ffffff",
        description:
          "Curabitur gravida nisi at nibh. In hac habitasse platea dictumst. Aliquam augue quam, sollicitudin vitae, consectetuer eget, rutrum at, lorem.\n\nInteger tincidunt ante vel ipsum. Praesent blandit lacinia erat. Vestibulum sed magna at nunc commodo placerat.\n\nPraesent blandit. Nam nulla. Integer pede justo, lacinia eget, tincidunt eget, tempus vel, pede.",
        price: 9.5,
        slug: "love-is-eternal-while-it-lasts",
        title_lower: "love is eternal while it lasts"
      },
      {
        id: "23ad0474-8b8d-420d-a2cc-dd54aa6ed589",
        title: "Lost World, The",
        genre: ["romance"],
        image: "http://dummyimage.com/100x100.png/cc0000/ffffff",
        description:
          "In congue. Etiam justo. Etiam pretium iaculis justo.\n\nIn hac habitasse platea dictumst. Etiam faucibus cursus urna. Ut tellus.\n\nNulla ut erat id mauris vulputate elementum. Nullam varius. Nulla facilisi.\n\nCras non velit nec nisi vulputate nonummy. Maecenas tincidunt lacus at velit. Vivamus vel nulla eget eros elementum pellentesque.\n\nQuisque porta volutpat erat. Quisque erat eros, viverra eget, congue eget, semper rutrum, nulla. Nunc purus.",
        price: 7.4,
        slug: "lost-world-the",
        title_lower: "lost world, the"
      }
    ]
  },
  cart: {
    dict: {
      "0b7b8b05-6c88-4517-a969-adf89a46bc0a": 1,
      "23ad0474-8b8d-420d-a2cc-dd54aa6ed589": 2
    },
    saved: false
  },
  filters: {
    string: "Lo",
    string_lower: "lo"
  },
  ui: {
    language: "en",
    version: 2,
    theme: "light"
  },
  routing: {
    locationBeforeTransitions: {
      pathname: "/genre/comedy",
      search: "",
      hash: "",
      key: "sb1up4"
    }
  },
  form: {}
};
