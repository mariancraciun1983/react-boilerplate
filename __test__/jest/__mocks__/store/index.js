import clone from 'clone-deep';

import * as FullData from '../../../../__mock__/api';

const defaultStore = {
  app: {
    state: 'loading',
  },
  auth: {
    authenticated: false,
    token: null,
    user: null,
  },
  cart: {
    dict: {},
    saved: true,
  },
  filters: {
    string: '',
    string_lower: '',
  },
  genres: {
    list: null,
  },
  movies: {
    list: null,
  },
  ui: {
    language: 'en',
    version: 0,
    theme: 'light',
  },
  routing: {},
  form: {},
};

const fullStore = clone(defaultStore);
fullStore.state = 'loaded';
fullStore.genres.list = FullData.genres;
fullStore.movies.list = FullData.movies.slice(0,200);
export default {
  defaultStore,
  fullStore,
}
