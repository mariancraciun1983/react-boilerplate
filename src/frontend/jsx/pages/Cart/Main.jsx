import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { push } from 'react-router-redux';
import {
  cartItems, cartTotal, cartSaved, authIsAuthenticated,
} from '../../../core/selectors';
import { updateMovie, save } from '../../../core/stores/reducers/cart';
import { formatPrice } from '../../../core/utils/formatter';
import Item from './components/Item';

const stateToProps = (state) => ({
  saved: cartSaved(state),
  items: cartItems(state),
  total: cartTotal(state),
  isAuthenticated: authIsAuthenticated(state),
});
const actionsToProps = (dispatch) => ({
  updateCartMovie: (id, qty) => dispatch(updateMovie(id, qty)),
  saveCart: () => dispatch(save()),
  redirect: (to) => dispatch(push(to)),
});
class Main extends React.PureComponent {
  constructor(props) {
    super(props);
    this.onUpdateMovie = this.onUpdateMovie.bind(this);
    this.onSave = this.onSave.bind(this);
  }

  onUpdateMovie(id, quantity) {
    const { updateCartMovie } = this.props;
    updateCartMovie(id, quantity);
  }

  onSave() {
    const { saveCart, isAuthenticated, redirect } = this.props;
    if (isAuthenticated) saveCart();
    else redirect('/auth/login');
  }

  render() {
    const { items, total, saved } = this.props;
    return (
      <>
        <div className="container">
          <div className="card-header">
            <h1>Shopping Cart</h1>
          </div>
          <div className="card-body">
            {items.map((item) => (
              <Item
                movie={item.movie}
                quantity={item.quantity}
                key={item.movie.id}
                onUpdate={this.onUpdateMovie}
              />
            ))}
            <div className="d-flex">
              <div className="flex-grow-1">
                <strong>{`Total: ${formatPrice(total)}$`}</strong>
              </div>
              <button
                type="submit"
                disabled={items.length === 0 || saved}
                onClick={this.onSave}
                className="btn btn-success btn-lg"
              >
                Save Cart
              </button>
            </div>
          </div>
        </div>
      </>
    );
  }
}

Main.propTypes = {
  updateCartMovie: PropTypes.func.isRequired,
  saveCart: PropTypes.func.isRequired,
  redirect: PropTypes.func.isRequired,
  isAuthenticated: PropTypes.bool.isRequired,
  saved: PropTypes.bool.isRequired,
  items: PropTypes.arrayOf(PropTypes.object).isRequired,
  total: PropTypes.number.isRequired,
};

export const Component = Main;
export default connect(
  stateToProps,
  actionsToProps,
)(Main);
