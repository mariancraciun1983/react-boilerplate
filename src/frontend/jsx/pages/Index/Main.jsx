import React from 'react';

import Nav from './components/Nav';
import Movies from './components/Movies';

class Index extends React.PureComponent {
  render() {
    return (
      <div className="d-flex">
        <div className="left-nav">
          <Nav />
        </div>
        <div className="flex-grow-1">
          <Movies />
        </div>
      </div>
    );
  }
}

export const Component = Index;
export default Index;
