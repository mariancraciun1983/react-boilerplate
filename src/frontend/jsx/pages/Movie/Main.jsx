import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { push } from 'react-router-redux';

import { movieById, genresBySlugs } from '../../../core/selectors';
import { updateMovie } from '../../../core/stores/reducers/cart';

const stateToProps = (state, props) => {
  // eslint-disable-next-line camelcase
  const { movie_id } = props.match.params;
  const movie = movieById(state, { movie_id });
  if (!movie) return {};
  const genres = genresBySlugs(state, movie.genre);
  return {
    movie,
    genres,
  };
};
const actionsToProps = (dispatch) => ({
  addToCart: (id) => dispatch(updateMovie(id, 1)),
  redirect: (to) => dispatch(push(to)),
});
class Main extends React.PureComponent {
  constructor(props) {
    super(props);
    this.addToCart = this.addToCart.bind(this);
  }

  componentDidMount() {
    const { movie } = this.props;
    if (!movie) {
      const { redirect } = this.props;
      redirect('/error/404');
    }
  }

  addToCart() {
    const { addToCart, movie } = this.props;
    addToCart(movie.id);
  }

  render() {
    const { movie, genres } = this.props;
    if (!movie) {
      return null;
    }
    return (
      <div className="container">
        <div className="card-header">
          <h1>{movie.title}</h1>
        </div>
        <div className="card-body">
          <div className="row">
            <div className="col col-5">
              <img src={movie.image} className="img-fluid w-100" alt="Movie" />
            </div>
            <div className="col col-7">
              <div>{movie.description}</div>
              <div>
                {genres.map((genre) => (
                  <span key={genre.slug}>
                    <Link to={`/genre/${genre.slug}`}>{genre.name}</Link>
                    &nbsp;
                  </span>
                ))}
              </div>
              <div className="d-flex mt-3">
                <div className="fz-30 pt-1 mr-3">
                  {`${movie.price}$`}
                </div>
                <button type="button" className="btn btn-lg btn-primary" onClick={this.addToCart}>
                  Add to Cart
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

Main.propTypes = {
  redirect: PropTypes.func.isRequired,
  addToCart: PropTypes.func.isRequired,
  movie: PropTypes.shape({
    id: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    image: PropTypes.string.isRequired,
    price: PropTypes.number.isRequired,
    description: PropTypes.string.isRequired,
  }),
  genres: PropTypes.arrayOf(PropTypes.object).isRequired,
};
Main.defaultProps = {
  movie: null,
};

export const Component = Main;
export default connect(
  stateToProps,
  actionsToProps,
)(Main);
