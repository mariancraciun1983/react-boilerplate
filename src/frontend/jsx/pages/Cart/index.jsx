import React from 'react';
import { Switch, Route, Redirect } from 'react-router';

import Main from './Main';

const routes = () => (
  <Switch>
    <Route component={Main} path="/cart" exact />
    <Redirect to={{ state: { error404: true, sourceModule: 'cart' } }} />
  </Switch>
);

export default { routes };
