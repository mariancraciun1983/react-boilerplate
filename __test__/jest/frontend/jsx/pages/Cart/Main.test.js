/// <reference types="jest" />
import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import {Component} from '../../../../../../src/frontend/jsx/pages/Cart/Main';
import { genericWrapper } from '../../../../__mocks__/componentWrappers';

describe("components/pages/Cart/Main", () => {
  let props;
  beforeEach(() => {
    props =  {
      updateCartMovie: jest.fn(),
      saveCart: jest.fn(),
      redirect: jest.fn(),
      saved: true,
      items: [],
      total: 10,
      isAuthenticated: true,
    }
  });

  it("Renders the cart", () => {
    expect.assertions(1);
    const tree = renderer
    .create(genericWrapper(<Component {...props} />))
    .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("Handles cart updates", () => {
    expect.assertions(1);
    const wrapper  = shallow(<Component {...props} />)
    wrapper.instance().onUpdateMovie('PROD_ID', 2);
    expect(props.updateCartMovie).toHaveBeenCalledWith('PROD_ID', 2);
  });

  it("Handles cart submission", () => {
    expect.assertions(3);
    let wrapper  = shallow(<Component {...props} />)
    wrapper.instance().onSave();
    expect(props.saveCart).toHaveBeenCalled();

    wrapper  = shallow(<Component {...props} isAuthenticated={false} />)
    wrapper.instance().onSave();
    expect(props.saveCart).toHaveBeenCalledTimes(1);//previouly
    expect(props.redirect).toHaveBeenCalled();
  });

});
