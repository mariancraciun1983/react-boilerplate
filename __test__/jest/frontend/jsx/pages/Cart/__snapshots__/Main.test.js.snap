// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/pages/Cart/Main Renders the cart 1`] = `
<div
  className="container"
>
  <div
    className="card-header"
  >
    <h1>
      Shopping Cart
    </h1>
  </div>
  <div
    className="card-body"
  >
    <div
      className="d-flex"
    >
      <div
        className="flex-grow-1"
      >
        <strong>
          Total: 10.00$
        </strong>
      </div>
      <button
        className="btn btn-success btn-lg"
        disabled={true}
        onClick={[Function]}
        type="submit"
      >
        Save Cart
      </button>
    </div>
  </div>
</div>
`;
